src/extensions/extensionsHelper.ts

```diff
@@ -503,52 +531,284 @@ export async function publishExtensionVersionFromLocalSource(args: {
+    if (response.ok) {
+      await response.body.pipe(unzipper.Extract({ path: tempDirectory.name })).promise(); // eslint-disable-line new-cap
+    }
+  } catch (err: any) {
     throw new FirebaseError(
-      `The version you are trying to publish (${clc.bold(
-        extensionSpec.version
-      )}) is lower than the current version (${clc.bold(
-        extension.latestVersion
-      )}) for the extension '${clc.bold(
-        `${args.publisherId}/${args.extensionId}`
-      )}'. Please make sure this version is greater than the current version (${clc.bold(
-        extension.latestVersion
-      )}) inside of extension.yaml.\n`,
-      { exit: 104 }
+      `Failed to fetch Extension archive ${archiveUri}. Please check the repo URI and source ref. ${err}`
     );
-  } else if (
-    extension &&
-    extension.latestVersion &&
-    semver.eq(extensionSpec.version, extension.latestVersion)
-  ) {
-    // publisher's version is equal to the current latest version.
+  }
+  const archiveName = fs.readdirSync(tempDirectory.name)[0];
+  const rootDirectory = path.join(tempDirectory.name, archiveName, extensionRoot);
+  // Pre-validation to show a more useful error message in the context of a temp directory.
+  try {
+    readFile(path.resolve(rootDirectory, EXTENSIONS_SPEC_FILE));
+  } catch (err: any) {
     throw new FirebaseError(
-      `The version you are trying to publish (${clc.bold(
-        extensionSpec.version
-      )}) already exists for the extension '${clc.bold(
-        `${args.publisherId}/${args.extensionId}`
-      )}'. Please increment the version inside of extension.yaml.\n`,
-      { exit: 103 }
+      `Failed to find ${clc.bold(EXTENSIONS_SPEC_FILE)} in directory ${clc.bold(
+        extensionRoot
+      )}. Please verify the root and try again.`
     );
   }
+  const { extensionSpec, notes } = await validateExtensionSpec({
+    publisherId: args.publisherId,
+    extensionId: args.extensionId,
+    rootDirectory: rootDirectory,
+    latestVersion: extension?.latestVersion,
+    stage: stage,
+  });
+
+  const sourceUri = path.join(repoUri, "tree", sourceRef, extensionRoot);
+  displayReleaseNotes(extensionRef, extensionSpec.version, notes, sourceUri);
+  const confirmed = await confirm({
+    nonInteractive: args.nonInteractive,
+    force: args.force,
+    default: false,
+  });
+  if (!confirmed) {
+    return;
+  }
+
+  // Publish the Extension version.
+  const extensionVersionRef = `${extensionRef}@${extensionSpec.version}`;
+  const publishSpinner = ora(`Publishing ${clc.bold(extensionVersionRef)}`);
+  let res;
+  try {
+    publishSpinner.start();
+    res = await publishExtensionVersion(
+      extensionVersionRef,
+      "",
+      extensionRoot,
+      repoUri,
+      args.sourceRef
+    );
+    publishSpinner.succeed(` Successfully published ${clc.bold(extensionRef)}`);
+  } catch (err: any) {
+    publishSpinner.fail();
+    if (err.status === 404) {
+      throw new FirebaseError(
+        marked(
```

Comment:
This string is long and repeated in publishFromLocalSource - separate it out into a constant?